src/packet/decode.ts

```diff
@@ -0,0 +1,96 @@
+import RLP = require("rlp");
+import {
+  IAuthMessagePacket,
+  IMessagePacket,
+  IWhoAreYouPacket,
+  Magic,
+  Packet,
+  Tag,
+} from "./types";
+
+import {
+  AUTH_TAG_LENGTH,
+  ERR_INVALID_BYTE_SIZE,
+  ERR_TOO_SMALL,
+  ERR_UNKNOWN_FORMAT,
+  ID_NONCE_LENGTH,
+  TAG_LENGTH,
+} from "./constants";
+
+/**
+ * Decode raw bytes into a packet. The `magic` value (SHA2256(node-id, b"WHOAREYOU")) is passed as a parameter to check
+ * for the magic byte sequence.
+ */
+export function decode(data: Buffer, magic: Magic): Packet {
```

Comment:
I think we should use "buffer" dependency and import {Buffer} from "buffer" here

Reply:
I don't like doing that because its not an optional polyfill. It will be slowing down / effecting nodejs by _never_ using the native Buffer methods and always using this browser-compatible implementation.

Reply:
Are you aware of any benchmarks ?

Reply:
https://github.com/feross/buffer#node-01114

Only concat is notably slower, not sure if it's gonna make that much difference. I don't think forcing everyone to use polyfill is a good idea, might be better to use buffer dep until we create lib to use native buffer if available

Reply:
Those are for a very old version of node. Recent benchmarks are different.

```
yarn run v1.12.3
$ node perf/bracket-notation.js && node perf/concat.js && node perf/copy-big.js && node perf/copy.js && node perf/new-big.js && node perf/new.js && node perf/readDoubleBE.js && node perf/readFloatBE.js && node perf/readUInt32LE.js && node perf/slice.js && node perf/writeFloatBE.js
BrowserBuffer#bracket-notation x 23,381,205 ops/sec ±0.41% (93 runs sampled)
Uint8Array#bracket-notation x 20,655,152 ops/sec ±3.06% (83 runs sampled)
NodeBuffer#bracket-notation x 21,969,275 ops/sec ±1.18% (86 runs sampled)
Fastest is BrowserBuffer#bracket-notation
BrowserBuffer#concat x 285,931 ops/sec ±2.32% (78 runs sampled)
Uint8Array#concat x 1,382,447 ops/sec ±3.31% (69 runs sampled)
NodeBuffer#concat x 3,344,152 ops/sec ±2.70% (79 runs sampled)
Fastest is NodeBuffer#concat
BrowserBuffer#copy(16000) x 1,248,823 ops/sec ±2.99% (80 runs sampled)
Uint8Array#copy(16000) x 2,693,750 ops/sec ±1.52% (88 runs sampled)
NodeBuffer#copy(16000) x 3,741,110 ops/sec ±0.66% (95 runs sampled)
Fastest is NodeBuffer#copy(16000)
BrowserBuffer#copy(16) x 3,245,863 ops/sec ±1.38% (91 runs sampled)
Uint8Array#copy(16) x 57,408,131 ops/sec ±2.37% (87 runs sampled)
NodeBuffer#copy(16) x 45,129,937 ops/sec ±1.51% (87 runs sampled)
Fastest is Uint8Array#copy(16)
BrowserBuffer#new(16000) x 394,361 ops/sec ±1.28% (83 runs sampled)
Uint8Array#new(16000) x 436,456 ops/sec ±1.87% (66 runs sampled)
NodeBuffer#new(16000) x 405,436 ops/sec ±1.43% (77 runs sampled)
Fastest is Uint8Array#new(16000)
BrowserBuffer#new(16) x 6,672,844 ops/sec ±2.66% (75 runs sampled)
Uint8Array#new(16) x 20,552,147 ops/sec ±1.83% (84 runs sampled)
NodeBuffer#new(16) x 13,772,807 ops/sec ±2.53% (78 runs sampled)
Fastest is Uint8Array#new(16)
BrowserBuffer#readDoubleBE x 308,759 ops/sec ±1.50% (90 runs sampled)
DataView#getFloat64 x 6,504,275 ops/sec ±0.65% (94 runs sampled)
NodeBuffer#readDoubleBE x 1,542,978 ops/sec ±0.59% (89 runs sampled)
Fastest is DataView#getFloat64
BrowserBuffer#readFloatBE x 681,618 ops/sec ±0.69% (90 runs sampled)
DataView#getFloat32 x 6,551,038 ops/sec ±0.83% (91 runs sampled)
NodeBuffer#readFloatBE x 2,737,416 ops/sec ±1.03% (90 runs sampled)
Fastest is DataView#getFloat32
BrowserBuffer#readUInt32LE x 2,555,711 ops/sec ±0.64% (88 runs sampled)
DataView#getUint32 x 6,535,575 ops/sec ±1.16% (92 runs sampled)
NodeBuffer#readUInt32LE x 5,367,406 ops/sec ±1.96% (88 runs sampled)
Fastest is DataView#getUint32
BrowserBuffer#slice x 3,177,345 ops/sec ±2.41% (85 runs sampled)
Uint8Array#subarray x 26,139,014 ops/sec ±2.61% (80 runs sampled)
NodeBuffer#slice x 20,485,810 ops/sec ±2.08% (80 runs sampled)
Fastest is Uint8Array#subarray
BrowserBuffer#writeFloatBE x 23,923 ops/sec ±0.86% (93 runs sampled)
DataView#setFloat32 x 3,657,511 ops/sec ±1.04% (89 runs sampled)
NodeBuffer#writeFloatBE x 2,376,096 ops/sec ±0.66% (94 runs sampled)
Fastest is DataView#setFloat32
Done in 199.80s.
```